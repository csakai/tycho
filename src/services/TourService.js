import Cookies from 'js-cookie';

export default class TourService {

  /**
   * Determines if the user has opted, via cookie, to skip the tour.
   *
   * @returns {Boolean} ability to skip tour
   */
  static canSkip = () => {
    return Cookies.get('tourViewed') === 'true';
  }

  /**
   * Sets a cookie to indicate that the tour was either viewed or skipped.
   */
  static setSkip = () => {
    Cookies.set('tourViewed', 'true', {expires: 365});
  }

  /**
   * Calculates the total time that showing labels will take,
   * separated and padded by the SEPARATION_INTERVAL constant.
   *
   * @param {Object[]} labels - list of labels to sum
   * @returns {Number} total duration
   */
  static getTourDuration = (labels) => {
    const SEPARATION_INTERVAL = 1000;

    let duration = SEPARATION_INTERVAL;
    
    duration += labels.reduce((cur, next) => {
      return cur + next.duration + SEPARATION_INTERVAL;
    }, 0);
    duration += SEPARATION_INTERVAL;

    return duration;
  }
}
