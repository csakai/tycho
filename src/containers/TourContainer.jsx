import React from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import * as UIControlsActions from '../actions/UIControlsActions';
import * as TourActions from '../actions/TourActions';
import * as LabelActions from '../actions/LabelActions';
import ReduxService from '../services/ReduxService';
import TourService from '../services/TourService';
import TourLabelContainer from './TourLabelContainer';
import Tour from '../components/Tour';
import Constants from '../constants';

export class TourContainer extends React.Component {

  static propTypes = {
      labels: PropTypes.array
  }

  componentDidMount = () => {
      if (TourService.canSkip()) {
          this.props.action.tourSkipped(true);
      }
  }

  componentWillReceiveProps = (nextProps) => {
      this.maybeSkipTour(nextProps);
      this.maybeStartTour(nextProps);
  }

  /**
   * Skips tour if opted to do so and is not already skipped.
   * 
   * @param {Object} nextProps
   * @param {Boolean} nextProps.isSkipped
   */
  maybeSkipTour = ({isSkipped}) => {
      if (this.props.isSkipped !== isSkipped && isSkipped) {
          this.skipTour();
      }
  }

  /**
   * Starts tour if not already initialized.
   * 
   * @param {Object} nextProps
   * @param {Boolean} nextProps.playing
   */
  maybeStartTour = ({playing}) => {
      if (playing && !this.props.playing && !this.props.isComplete) {
          this.initializeTour();
      }
  }

  /**
   * Returns true if the tour should be run.
   * 
   * @returns {Boolean}
   */
  shouldRunTour = () => {
      return this.props.playing && !this.props.isSkipped;
  }

  /**
   * Initializes the tour.
   */
  initializeTour = () => {
      const {action, labels} = this.props;
      const tourDuration = TourService.getTourDuration(labels);

      if (!TourService.canSkip()) {
          action.setUIControls(false);
          action.setCameraOrbit(true);

          setTimeout(this.onOrbitComplete, tourDuration);
      }
  }
  
  /**
   * Callback to invoke once the tour orbit has completed.
   */
  onOrbitComplete = () => {
      if (!this.props.isComplete) {
          this.setDefaultActiveOrbital();
          setTimeout(this.onTourComplete, Constants.WebGL.Tween.SLOW);
      }
  }

  /**
   * Callback to invoke once the entire tour is completed.
   */
  onTourComplete = () => {
      this.props.action.tourCompleted(true);
  }

  /**
   * Skips the tour and zooms directly into the target planet.
   */
  skipTour = () => {
      const {action} = this.props;

      action.tourCompleted(true);
      action.setCameraOrbit(false);
      action.setUIControls(true);

      this.setDefaultActiveOrbital();
  }

  /**
   * Sets the active orbital targetId and the header label text to UI defaults.
   */
  setDefaultActiveOrbital = () => {
      this.props.action.setActiveOrbital(Constants.UI.ALTERNATE_TARGET_NAME);
      this.props.action.setActiveOrbital(Constants.UI.DEFAULT_TARGET_NAME);
      this.props.action.setLabelText(Constants.UI.DEFAULT_LABEL_TEXT);
  }

  /**
   * Sets the skip tour cookie and skips tour.
   */
  skipTourTrigger = () => {
      TourService.setSkip();
      this.skipTour();
  }
  
  /**
   * Maps the labels list prop to TourLabelContainers.
   *
   * @param {Object[]} labels - list of labels
   * @returns {TourLabelContainer[]}
   */
  getLabels = (labels) => {
      const separation = Constants.Tour.SEPARATION_INTERVAL;
      let totalTime = separation;
    
      return labels.map(({text, duration}, key) => {
          totalTime += separation;
          const start = totalTime;
          totalTime += duration;
          const end = totalTime;

          return (
              <TourLabelContainer
                  key={key}
                  text={text}
                  start={start}
                  end={end}
              />
          );
      });
  }

  render() {
      if (!this.shouldRunTour()) {
          return null;
      }
      return (
          <Tour
              {...this.props}
              skipTour={this.skipTourTrigger}
              labels={this.getLabels(this.props.labels)}
          />
      );
  }
}

export default connect(
    ReduxService.mapStateToProps(
        'uiControls.controlsEnabled',
        'uiControls.scale',
        'label.targetId',
        'tour.isComplete',
        'tour.isSkipped',
        'animation.playing',
        'data.pageText'
    ),
    ReduxService.mapDispatchToProps(
        UIControlsActions,
        TourActions,
        LabelActions
    )
)(TourContainer);