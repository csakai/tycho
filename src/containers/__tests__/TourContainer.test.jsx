import React from 'react';
import toJson from 'enzyme-to-json';
import {shallow} from 'enzyme';
import {TourContainer} from '../TourContainer';
import TourService from '../../services/TourService';
import Constants from '../../constants';

const labels = [
    {
        duration: 5000,
        text: 'Welcome to the Solar System'
    },
    {
        duration: 5000,
        text: 'This is a real-time interactive simulation of major planetary bodies'
    },
    {
        duration: 3000,
        text: 'Let\'s start exploring'
    }
];

jest.useFakeTimers();

describe('Tour Container', () => {
    let component;
    let tourContainer;
    let action;

    beforeEach(() => {
        action = {
            setUIControls: jest.fn(),
            setCameraOrbit: jest.fn(),
            setActiveOrbital: jest.fn(),
            setLabelText: jest.fn(),
            tourCompleted: jest.fn(),
            tourSkipped: jest.fn(),
            showLabel: jest.fn()
        };

        component = shallow(
            <TourContainer
                labels={labels}
                action={action}
                pageText={{}}
            />
        );
        tourContainer = component.instance();
    });

    describe('componentDidMount()', () => {
        let spy;

        beforeEach(() => {
            spy = jest.spyOn(tourContainer.props.action, 'tourSkipped');
        });

        it('should call the tour skip action if the tour can be skipped', () => {
            TourService.canSkip = () => true;
            tourContainer.componentDidMount();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
            expect(spy).toHaveBeenCalledWith(true);
        });

        it('should not call the tour skip action if the tour cannot be skipped', () => {
            TourService.canSkip = () => false;
            tourContainer.componentDidMount();

            expect(spy).not.toHaveBeenCalled();
        });
    });

    describe('componentWillReceiveProps()', () => {
        beforeEach(() => {
            tourContainer.maybeSkipTour = jest.fn();
            tourContainer.maybeStartTour = jest.fn();
        });

        it('should call maybeSkipTour() with nextProps', () => {
            const spy = jest.spyOn(tourContainer, 'maybeSkipTour');
            const nextProps = {isSkipped: true};

            tourContainer.componentWillReceiveProps(nextProps);

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
            expect(spy).toHaveBeenCalledWith(nextProps);
        });

        it('should call maybeStartTour() with nextProps', () => {
            const spy = jest.spyOn(tourContainer, 'maybeStartTour');
            const nextProps = {playing: true};

            tourContainer.componentWillReceiveProps(nextProps);

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
            expect(spy).toHaveBeenCalledWith(nextProps);
        });
    });

    describe('maybeSkipTour()', () => {
        let spy;

        beforeEach(() => {
            spy = jest.spyOn(tourContainer, 'skipTour');
        });

        it('should call skipTour() if the isSkipped prop has changed and is true', () => {
            tourContainer.maybeSkipTour({isSkipped: true});

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
        });

        it('should not call skipTour() if the isSkipped prop is false', () => {
            tourContainer.maybeSkipTour({isSkipped: false});
            expect(spy).not.toHaveBeenCalled();
        });

        it('should not call skipTour() if the isSkipped prop has not changed', () => {
            tourContainer.maybeSkipTour({});
            expect(spy).not.toHaveBeenCalled();
        });
    });

    describe('maybeStartTour()', () => {
        beforeEach(() => {
            tourContainer.initializeTour = jest.fn();
        });

        it('should initialize the tour if the user-entered status has changed to true', () => {
            const spy = jest.spyOn(tourContainer, 'initializeTour');

            tourContainer.props = {playing: false};
            tourContainer.maybeStartTour({playing: true});

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
        });

        it('should initialize the tour if the user-entered status is false', () => {
            const spy = jest.spyOn(tourContainer, 'initializeTour');

            tourContainer.maybeStartTour({playing: false});

            expect(spy).not.toHaveBeenCalled();
        });

        it('should not initialize the tour if it has already been initialized', () => {
            const spy = jest.spyOn(tourContainer, 'initializeTour');

            tourContainer.props = {playing: true};
            tourContainer.maybeStartTour({playing: true});

            expect(spy).not.toHaveBeenCalled();
        });
    });

    describe('shouldRunTour()', () => {
        describe('when user has entered scene', () => {
            beforeEach(() => {
                tourContainer.props = {
                    playing: true
                };
            });

            it('should return true when not skipped', () => {
                tourContainer.props.isSkipped = false;
                expect(tourContainer.shouldRunTour()).toBe(true);
            });

            it('should return false when skipped', () => {
                tourContainer.props.isSkipped = true;
                expect(tourContainer.shouldRunTour()).toBe(false);
            });
        });

        describe('when user has not yet entered scene', () => {
            it('should return false', () => {
                tourContainer.props = {
                    playing: false,
                    isSkipped: false
                };
                expect(tourContainer.shouldRunTour()).toBe(false);
            });
        });
    });

    describe('initializeTour()', () => {
        beforeEach(() => {
            tourContainer.getTourDuration = jest.fn();
            TourService.canSkip = () => false;
        });

        describe('when the tour cannot be skipped', () => {
            it('should call setUIControls with false', () => {
                const spy = jest.spyOn(tourContainer.props.action, 'setUIControls');

                tourContainer.initializeTour();

                expect(spy).toHaveBeenCalled();
                expect(spy).toHaveBeenCalledTimes(1);
                expect(spy).toHaveBeenCalledWith(false);
            });

            it('should call setCameraOrbit with true', () => {
                const spy = jest.spyOn(tourContainer.props.action, 'setCameraOrbit');

                tourContainer.initializeTour();

                expect(spy).toHaveBeenCalled();
                expect(spy).toHaveBeenCalledTimes(1);
                expect(spy).toHaveBeenCalledWith(true);
            });

            it('should call onOrbitComplete after the calculated duration has passed', () => {
                tourContainer.onOrbitComplete = jest.fn();
                const spy = jest.spyOn(tourContainer, 'onOrbitComplete');

                tourContainer.initializeTour();

                expect(spy).not.toHaveBeenCalled();

                jest.runAllTimers();

                expect(spy).toHaveBeenCalled();
                expect(spy).toHaveBeenCalledTimes(1);
            });
        });

        describe('when the tour can be skipped', () => {
            beforeEach(() => {
                TourService.canSkip = () => true;
            });

            it('should not call setCameraOrbit with true', () => {
                const spy = jest.spyOn(tourContainer.props.action, 'setCameraOrbit');

                tourContainer.initializeTour();
                expect(spy).not.toHaveBeenCalled();
            });
        });
    });

    describe('onOrbitComplete()', () => {
        beforeEach(() => {
            tourContainer.setDefaultActiveOrbital = jest.fn();
        });

        it('should call setActiveOrbital with the default orbital name if tour is not completed', () => {
            const spy = jest.spyOn(tourContainer, 'setDefaultActiveOrbital');

            tourContainer.props = {isComplete: false};
            tourContainer.onOrbitComplete();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
        });

        it('should not call setActiveOrbital if tour is completed', () => {
            const spy = jest.spyOn(tourContainer, 'setDefaultActiveOrbital');

            tourContainer.props = {isComplete: true};
            tourContainer.onOrbitComplete();

            expect(spy).not.toHaveBeenCalled();
        });
    });

    describe('onTourComplete()', () => {
        it('should call tourCompleted with true', () => {
            const spy = jest.spyOn(tourContainer.props.action, 'tourCompleted');

            tourContainer.onTourComplete();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
            expect(spy).toHaveBeenCalledWith(true);
        });
    });

    describe('skipTour()', () => {
        it('should call the tourCompleted action with `true`', () => {
            const spy = jest.spyOn(tourContainer.props.action, 'tourCompleted');

            tourContainer.skipTour();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
            expect(spy).toHaveBeenCalledWith(true);
        });

        it('should call the setCameraOrbit action with `false`', () => {
            const spy = jest.spyOn(tourContainer.props.action, 'setCameraOrbit');

            tourContainer.skipTour();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
            expect(spy).toHaveBeenCalledWith(false);
        });

        it('should call setDefaultActiveOrbital()', () => {
            const spy = jest.spyOn(tourContainer, 'setDefaultActiveOrbital');

            tourContainer.skipTour();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
        });
    });

    describe('setDefaultActiveOrbital()', () => {
        it('should call the setActiveOrbital action with the alternate and default orbitals', () => {
            const spy = jest.spyOn(tourContainer.props.action, 'setActiveOrbital');

            tourContainer.setDefaultActiveOrbital();

            setTimeout(() => {
                expect(spy).toHaveBeenCalled();
                expect(spy).toHaveBeenCalledTimes(2);
                expect(spy).toHaveBeenCalledWith(...Constants.UI.Targets.ALTERNATE);
                expect(spy).toHaveBeenCalledWith(...Constants.UI.Targets.DEFAULT);
            });
        });
    });

    describe('skipTourTrigger()', () => {
        it('should call skipTour()', () => {
            const spy = jest.spyOn(tourContainer, 'skipTour');

            tourContainer.skipTourTrigger();

            expect(spy).toHaveBeenCalled();
            expect(spy).toHaveBeenCalledTimes(1);
        });
    });

    describe('getLabels()', () => {
        it('should return an array of TourLabels', () => {
            const result = tourContainer.getLabels(labels);

            expect(Array.isArray(result)).toBe(true);

            result.forEach((label) => {
                expect(label).toMatchSnapshot();
            });
        });
    });

    describe('render()', () => {
        it('should return null when tour cannot be run', () => {
            tourContainer.shouldRunTour = () => false;
            tourContainer.render();

            expect(toJson(component)).toEqual(null);
        });

        it('should render the tour container successfully when tour can be run', () => {
            tourContainer.shouldRunTour = () => true;
            tourContainer.render();

            expect(toJson(component)).toMatchSnapshot();
        });
    });
});
