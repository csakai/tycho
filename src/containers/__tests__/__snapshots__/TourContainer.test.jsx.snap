// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tour Container getLabels() should return an array of TourLabels 1`] = `
<TourLabelContainer
  end={9000}
  start={4000}
  text="Welcome to the Solar System"
/>
`;

exports[`Tour Container getLabels() should return an array of TourLabels 2`] = `
<TourLabelContainer
  end={16000}
  start={11000}
  text="This is a real-time interactive simulation of major planetary bodies"
/>
`;

exports[`Tour Container getLabels() should return an array of TourLabels 3`] = `
<TourLabelContainer
  end={21000}
  start={18000}
  text="Let's start exploring"
/>
`;

exports[`Tour Container render() should render the tour container successfully when tour can be run 1`] = `null`;
